Save Scaled Background Subtraction Metadata to Subtracted Files

cailafinn opened the pull request:
### Description of work

#### Summary of work
When saving out background subtracted workspaces using the recently added (#35811) scaled background subtraction work, the name of the workspace used and the scale factor it was multiplied by are now saved to the NX and CanSAS1D output files. 

Fixes: #35939


**Report to:** @smk78 


#### Further detail of work
- Since the scaled background subtraction saves out the non-subtracted workspaces as well, these files do not include the additional metadata.

### To test:
1. In your Manage User Directories window, ensure that your save location is somewhere sensible. Ideally a new empty directory as it'll make finding the files you need much easier. 
2. Open the ISIS SANS Interface
3. Load the batch and user files from the loqdemo in the Training Course Data. 
4. Switch the reduction mode to HAB in the Settings tab and run the reduction. 
5. Take note of one of the output workspaces produced.
6. Make a copy of one of the lines currently in the runs table.
7. Give it a new output name. 
8. Tick the Scaled Background Subtraction checkbox
9. In your new row, enter in the name of the workspace in your workspace column
10. Enter a decimal value in the scale factor column (0.5 is what I generally use)
11. Check that the reduction mode is still HAB
12. Change the save options to output both NXCanSAS and CanSAS1D formats
13. Run the reduction for your new line
14. In your save directory, check that the values you entered in the table for scaled background workspace and scale factor are present in the process notes section of the output bgsub xml file. 
15. Use [HDFView](https://www.hdfgroup.org/downloads/hdfview/) to do the same for the `.h5` files. 

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

Commit:
Add saving of bgsub metadata if available

RE #35939

Commit:
Implement bgsub metadata parameters in low-level algorithms

RE #35939

Commit:
Move BGSub parameters to CanSAS1Dv2

RE #35939

Commit:
Save bgsub metadata into CanSAS1D files

RE #35939

Commit:
Fix using strings in place of floats

RE #35939

Commit:
Save bgsub info to NXCanSAS format

RE #35939

Commit:
Fix metadata being applied to non-subtracted files

RE #35939

Commit:
Test only saving subtracted workspaces with bgsub metadata

RE #35939

Commit:
Add consts to pass cppcheck

RE #35939

rbauststfc reviewed (commented):
The changes are working well - saving was successful and re-loading worked without error (we don't actually re-load the background subtraction information anywhere, which is expected).

I did notice that if I used the option name for the background workspace in the batch table rather than the full workspace name then that is what gets put into the file metadata. I wasn't sure if it should really be the full workspace name in the ADS?

On the code changes, they look good, I just had a couple of comments and queries. Also, are we missing releasing notes for this change?

cailafinn commented on Framework/DataHandling/inc/MantidDataHandling/NXcanSASDefinitions.h:
```diff
@@ -134,6 +134,8 @@ const std::string sasProcessTermSampleTrans = "sample_trans_run";
 const std::string sasProcessTermSampleDirect = "sample_direct_run";
 const std::string sasProcessTermCanScatter = "can_scatter_run";
 const std::string sasProcessTermCanDirect = "can_direct_run";
+const std::string sasProcessTermScaledBgSubWorkspace = "scaled_bgsub_workspace";
```
I'm more keen to keep consistency throughout the process. "Can" is usually used to represent a can-only run that is done in every individual reduction. Typically the can would be subtracted from these background workspaces as well (leaving only the background fluid scatter info), so I think it makes more sense to keep to the "Scaled Background Subtraction" name as is used on the GUI.

Commit:
Add bgsub suffix to constants file

RE #35939

Commit:
Fix batch row name not workspace name being saved to metadata

RE #35939

Commit:
Refactor adding of additional metadata

RE #35939

Commit:
Fix adjustment of name in state over package state

Each reduction package has it's own copy of the initial SANS State.
This is the one we should be editing so that the right metadata goes out
to the saved file.

RE #35939

Commit:
Add a release note

RE #35939

Commit:
Add tests for changes to save_to_file

RE #35959

rbauststfc commented on scripts/test/SANS/algorithm_detail/batch_execution_test.py:
```diff
@@ -335,22 +337,49 @@ def test_that_save_workspace_to_file_includes_run_numbers_in_options(self, mock_
-        mock_alg_manager.assert_called_once_with("SANSSave", **expected_options)
+    @mock.patch("sans.algorithm_detail.batch_execution.get_all_names_to_save")
+    @mock.patch("sans.algorithm_detail.batch_execution.save_workspace_to_file")
+    def test_that_subtracted_save_to_file_does_not_includes_metadata_in_options(self, mock_save_func, mock_names_func):
```
Was this test name intended to be: `test_that_subtracted_save_to_file_includes_metadata_in_options`?

rbauststfc reviewed (commented):
Code changes look good and well-structured for maintainability. Good test coverage and all the functional changes are working well, with the full name for the background subtraction workspace now outputting to the file.

There just looks like there might be a small typo in one of the test names, but otherwise this looks good to go in. 👍

Commit:
Fix typo

RE #35939

rbauststfc reviewed (approved):
Thanks for sorting the typo. This is all looking great now.
